redsky/event_streams.py

```diff
@@ -0,0 +1,711 @@
+    i : int
+        Counter for the number of events
+    run_start_uid : None
+    provenance : {}
+    event_failed : False
+    """
+    def __init__(self, child=None, children=None,
+                 *, output_info=None, input_info=None, md=None,
+                 **kwargs):
+        """Initialize the stream
+
+        Parameters
+        ----------
+        input_info: dict, optional
+            describes the incoming streams
+        output_info: list of tuples, optional
+            describes the resulting stream
+        md: dict, optional
+            Additional metadata to be added to the run start document
+
+        Notes
+        ------
+        input_info is designed to map keys in streams to kwargs in functions.
+        It is critical for the internal data from the events to be returned,
+        upon `event_guts`.
+        input_info = {'input_kwarg': ('data_key', stream_number)}
+        Note that the stream number is assumed to be zero if not specified
+
+        output_info is designed to take the output tuple and map it back into
+        data_keys.
+        output_info = [('data_key', {'dtype': 'array', 'source': 'testing'})]
+        """
+        Stream.__init__(self, child, children)
+        if md is None:
+            md = {}
+        if output_info is None:
+            output_info = {}
+        if input_info is None:
+            input_info = {}
+        self.outbound_descriptor_uid = None
+        self.md = md
+        self.output_info = output_info
+        self.input_info = input_info
+
+        # TODO: need multiple counters for multiple descriptors
+        # This will need to be a dict with keys of descriptor names
+        self.i = None
+        self.run_start_uid = None
+        self.provenance = {}
+        self.event_failed = False
+
+        # If the stream number is not specified its zero
+        for k, v in input_info.items():
+            if len(v) < 2 or isinstance(v, str):
+                input_info[k] = (v, 0)
+            if isinstance(v[1], Stream):
+                input_info[k] = (v[0], self.children.index(v[1]))
+
+    def emit(self, x):
+        """ Push data into the stream at this point
+
+        This is typically done only at source Streams but can theoretically be
+        done at any point
+        """
+        if x is not None:
+            result = []
+            for parent in self.parents:
+                r = parent.update(x, who=self)
+                if type(r) is list:
+                    result.extend(r)
+                else:
+                    result.append(r)
+            return [element for element in result if element is not None]
+
+    def dispatch(self, nds):
+        """Call method which correlates to the input document
+
+        Parameters
+        ----------
+        nds: tuple
+            Name document pair
+
+        Returns
+        -------
+        tuple:
+            New name document pair
+        """
+        name, docs = self.curate_streams(nds)
```

Comment:
yeah, line 602 of `event_streams.py` is where it's used in `bundle`. I didn't see anywhere else where `curate_streams` would do its magic (other than just pass through the single name doc pair). I'm curious to see how often it's used and where. (I didn't go through in complete detail which is why I'm asking as a shortcut)